cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -218,18 +220,21 @@ public Future<Void> rollingRestart(Function<Pod, RestartReasons> podNeedsRestart
         singleExecutor.submit(() -> {
             LOGGER.debugCr(reconciliation, "Verifying cluster pods are up-to-date.");
             List<NodeRef> pods = new ArrayList<>(nodes.size());
-
+            List<NodeRef> unReadyPods = new ArrayList<>();
+            List<NodeRef> readyPods = new ArrayList<>();
             for (NodeRef node : nodes)  {
                 // Order the nodes unready first otherwise repeated reconciliations might each restart a pod
                 // only for it not to become ready and thus drive the cluster to a worse state.
-
-                // TODO: In KRaft mode, We currently roll only nodes with the broker role. This is because of Kafka
-                //       limitations. Once managing controller nodes is supported with Kafka Admin API, this should be
-                //       fixed. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/8593.
-                if (node.broker()) {
-                    pods.add(podOperations.isReady(namespace, node.podName()) ? pods.size() : 0, node);
+                // in KRaft mode Roll controllers first, then brokers, but roll unready brokers before ready controllers
```

Comment:
"but roll" makes this sentence a little hard to parse. Better to just document the order as a list. 

I think there's a problem with rolling unhealthy brokers first: If the controller quorum is also not healthy then a restarted broker might not be able to register itself with the cluster (because doing so requires a metadata write which is impossible because the controller quorum is not working). 

Of the top of my head I'd suggest the following ordering:

1. Unhealthy controllers (active or not).
2. Healthy non-active controllers that require a restart.
3. Active controller that requires a restart (unless already restarted).
4. Unhealthy pure brokers.
5. Healthy pure brokers that require a restart.

Reply:
We went back and forth with the order, initially we were doing what you have suggested (unready controllers, ready controllers, unready brokers and ready brokers), which I think makes sense and happy to go with. 

The reason we changed the order was because we thought if we are rolling unready controllers first, that should bring the quorum to healthy state. If there are unready brokers, rolling controllers could bring the cluster to worse state so perhaps, rolling everything that are unready first to get the cluster to healthy state before progressing with any of the ready pods. 

One thing, we are missing in this PR is deferring the active controller so that we roll it last from controller list or full list of pods. At the beginning we were not sure, what's the best way to identify the active controller but now that we have implemented describeMetadataQuorum call to the Admin API, we can do that easily.

Reply:
We've changed the order to be as you suggested up above. The upshot of the change is that if we fail to roll the KRaft controllers, we don't try to roll the brokers at all. We've separated them into two different Future compositions.